tools/block-generator/generator/generate.go

```diff
@@ -759,6 +765,93 @@ func (g *generator) generateAssetTxnInternalHint(txType TxTypeID, round uint64,
 	return
 }
 
+// ---- 3. App Transactions ----
+
+func (g *generator) generateAppTxn(round uint64, intra uint64) (transactions.SignedTxn, transactions.ApplyData, error) {
+	start := time.Now()
+	selection, err := weightedSelection(g.appTxWeights, getAppTxOptions(), appSwapCall)
+	if err != nil {
+		return transactions.SignedTxn{}, transactions.ApplyData{}, err
+	}
+
+	actual, txn, err := g.generateAppCallInternal(selection.(TxTypeID), round, intra, 0, nil)
+	if err != nil {
+		return transactions.SignedTxn{}, transactions.ApplyData{}, fmt.Errorf("unexpected error received from generateAppCallInternal(): %w", err)
+	}
+	if txn.Type == "" {
+		return transactions.SignedTxn{}, transactions.ApplyData{}, fmt.Errorf("missing transaction type for app transaction")
+	}
+	defer g.recordData(actual, start)
+
+	return signTxn(txn), transactions.ApplyData{}, nil
+}
+
+func (g *generator) generateAppCallInternal(txType TxTypeID, round, intra, hintIndex uint64, hintApp *appData) (TxTypeID, transactions.Transaction, error) {
+	actual := txType
+
+	isApp, kind, appTx, err := parseAppTxType(txType)
+	if err != nil {
+		return "", transactions.Transaction{}, err
+	}
+	if !isApp {
+		return "", transactions.Transaction{}, fmt.Errorf("should be an app but not parsed that way: %v", txType)
+	}
+	if appTx != appTxTypeCreate {
+		return "", transactions.Transaction{}, fmt.Errorf("invalid transaction type forapp %v", appTx)
+	}
+
+	var senderIndex uint64
+	if hintApp != nil {
+		return "", transactions.Transaction{}, fmt.Errorf("not ready for hint app %v", hintApp)
+	} else {
+		senderIndex = rand.Uint64() % g.numAccounts
+	}
+
+	actualAppTx := appTx
+
+	numApps := uint64(len(g.apps[kind]))
+	if numApps == 0 {
+		actualAppTx = appTxTypeCreate
+	}
```

Comment:
is this block necessary? you've assert at line 799 that appTx must be appTxTypeCreate

Reply:
This is another one I'd like to keep till the next PR is ready, and so let's not resolve this discussion.

Reply:
Logic has been fixed in #5478